### PR TITLE
Exec probes should not be unbounded

### LINKED ISSUE #70890: Kubelet inmediatelly eats all node's memory when livenessProbe command outputs infinite stdout
**What happened**:
If a pod configures a livenessProbe command that just outputs infinitelly through stdout, kubelet quickly ends up eating all node's memory, possibly ending up with OOM kills and the node going to unhealthy state.

**What you expected to happen**:
Kubelet increases memory to store a certain amount of logs. The rest are just  dropped.

**How to reproduce it (as minimally and precisely as possible)**:
You just need to create the next pod:
```
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
spec:
  containers:
  - name: myapp-container
    image: busybox
    readinessProbe:
      exec:
        command:
        - "yes"
      failureThreshold: 3
      initialDelaySeconds: 10
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    livenessProbe:
      exec:
        command:
        - "yes"
      failureThreshold: 3
      initialDelaySeconds: 10
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    command: ['sh', '-c', 'echo Hello Kubernetes! && sleep 3600']
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):  Reproduced on 1.6, 1.9 and 1.12
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): CoreOS 1800.7.0
- Kernel (e.g. `uname -a`): Linux ip-10-24-15-116.eu-west-1.compute.internal 4.14.63-coreos #1 SMP Wed Aug 15 22:26:16 UTC 2018 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux
- Install tools: kube-aws (except for 1.12, which was using DCOS)
- Others:

/kind bug

### PR DESCRIPTION
**Pull request by dims**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
In 1f270ef4e29fe699300a67eb096b80590dfc877e, we added 10KB as the read
limit for http probes. we should do the same for exec probes as well.

Co-Authored-By: Tim Allclair <tallclair@google.com>

Change-Id: If154c5c4e669829ab94839c56260a894a6714f0f

**Which issue(s) this PR fixes**:

Fixes #70890

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Limit the body length of exec readiness/liveness probes. remote CRIs and Docker shim read a max of 16MB output of which the exec probe itself inspects 10kb.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by dims** [comment]
cc @tallclair

File: pkg/probe/exec/exec.go [hunk 1]
```diff
@@ -39,7 +44,34 @@ type execProber struct{}
...
+	go func() {
+		b, _ := utilio.ReadAtMost(stderrPipe, maxReadLength)
+		stderr <- string(b)
+	}()
+
+	if err := e.Start(); err != nil {
+		return probe.Failure, "", err
+	}
+
+	err = e.Wait()
```
Comment (liggitt) [inline pkg/probe/exec/exec.go]: will this block until output is consumed? Do we need to drain stdout/stderr?

Comment (dims) [reply pkg/probe/exec/exec.go]: good point, https://golang.org/pkg/os/exec/#Cmd.Wait says

```
If any of c.Stdin, c.Stdout or c.Stderr are not an *os.File, Wait also waits for the respective I/O loop copying to or from the process to complete.
```

Comment (dims) [reply pkg/probe/exec/exec.go]: gotta test more tomorrow

Comment (dims) [reply pkg/probe/exec/exec.go]: updated code and tested as well

File: pkg/kubelet/prober/prober.go [hunk 2]
```diff
@@ -298,17 +300,18 @@ func (eic execInContainer) Stop() {
...
 
 func (eic execInContainer) Wait() error {
-	return fmt.Errorf("unimplemented")
+	return nil
 }
 
 func (eic execInContainer) StdoutPipe() (io.ReadCloser, error) {
-	return nil, fmt.Errorf("unimplemented")
+	b, err := eic.run()
+	return ioutil.NopCloser(bytes.NewReader(b)), err
```
Comment (mattjmcnaughton) [inline pkg/kubelet/prober/prober.go]: Can you share more about the `NopCloser`? What makes it preferable here?

Comment (mattjmcnaughton) [reply pkg/kubelet/prober/prober.go]: Is that we are reading from `Stdout/Stderr`? So it doesn't actually want to close the file?

File: pkg/probe/exec/exec.go [hunk 3]
```diff
@@ -39,7 +46,52 @@ type execProber struct{}
...
+	if err := e.Start(); err != nil {
+		exit, ok := err.(exec.ExitError)
+		if ok && exit.ExitStatus() != 0 {
+			return probe.Failure, "", nil
+		}
+		return probe.Unknown, "", fmt.Errorf("error reading start : %s", err)
+	}
+
+	err = e.Wait()
+	data := <-stdout + <-stderr
```
Comment (mattjmcnaughton) [inline pkg/probe/exec/exec.go]: Confirming that we won't see execution hang at this point if there is either no stdout or no stderr?

**Review by mattjmcnaughton — COMMENTED** [review]
Overall, makes sense to me.

Confirming this is part of the overall k8s push to not allow unbounded input?

File: pkg/kubelet/prober/prober.go (hunk 2 above)
Comment (dims) [reply pkg/kubelet/prober/prober.go]: yep, we are readying from Stdout/Stderr

File: pkg/probe/exec/exec.go (hunk 3 above)
Comment (dims) [reply pkg/probe/exec/exec.go]: AFAICT. yes!

**Comment by dims** [comment]
/test pull-kubernetes-e2e-gce

failure unrelated to the PR change

File: pkg/probe/exec/exec.go [hunk 4]
```diff
@@ -39,7 +46,52 @@ type execProber struct{}
 // from executing a command. Returns the Result status, command output, and
 // errors if any.
 func (pr execProber) Probe(e exec.Cmd) (probe.Result, string, error) {
-	data, err := e.CombinedOutput()
+	stdoutPipe, err := e.StdoutPipe()
```
Comment (tallclair) [inline pkg/probe/exec/exec.go]: I started a PR for this with adding a LimitWriter utility: https://github.com/tallclair/kubernetes/commit/843ef55d715735d2672f020305ef4771ddaf55fd

I think you can just set `e.Stdout` and `e.Stderr` to a limit writer, and simplify a lot of the following logic.

Comment (tallclair) [reply pkg/probe/exec/exec.go]: Note that setting them to the same writer will interleave the output too. This is all that `CombinedOutput` is doing under the hood (https://golang.org/src/os/exec/exec.go?s=15900:15946#L541)

**Review by tallclair — COMMENTED** [review]
Under the hood, the exec probe is calling `RunInContainer` which is just a call to `ExecSync`: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container.go#L793

This is OK for "remote" (aka CRI) runtimes, since gRPC caps the response size at 16Mb by default, but dockershim is still vulnerable as it's in-process. To fix it, we need to set the docker call to exec to use a LimitWriter: https://github.com/kubernetes/kubernetes/blob/1888745c3db2432a613b2649686ee6bf5c9d2fa1/pkg/kubelet/dockershim/docker_streaming.go#L78-L85

If you want to pull my LimitWriter commit into this PR and fix dockershim, go for it.

File: pkg/kubelet/dockershim/docker_streaming.go [hunk 5]
```diff
@@ -78,8 +80,8 @@ func (ds *dockerService) ExecSync(_ context.Context, req *runtimeapi.ExecSyncReq
 	var stdoutBuffer, stderrBuffer bytes.Buffer
 	err := ds.streamingRuntime.exec(req.ContainerId, req.Cmd,
 		nil, // in
-		ioutils.WriteCloserWrapper(&stdoutBuffer),
-		ioutils.WriteCloserWrapper(&stderrBuffer),
+		ioutils.WriteCloserWrapper(ioutils.LimitWriter(&stdoutBuffer, maxMsgSize)),
```
Comment (dims) [inline pkg/kubelet/dockershim/docker_streaming.go]: @tallclair we cannot pass through a size parameter to `ExecSync` from `RunInContainer` as that would mean a change in signature of `ExecSync`. So i just set the LimitWriter to the same 16MB (that is the limit we set for GPRC for remote CRIs)

Comment (tallclair) [reply pkg/kubelet/dockershim/docker_streaming.go]: That makes sense to me.

**Comment by tallclair** [comment]
/lgtm
Have you had a chance to manually verify this?

**Comment by dims** [comment]
@tallclair manually testing right now while the CI is churning

**Comment by dims** [comment]
tested manually using the following patch as well (see liveness probe yaml below as well)

```
diff --git a/pkg/kubelet/dockershim/docker_streaming.go b/pkg/kubelet/dockershim/docker_streaming.go
index 76b2fecd11b8..2328b07fc121 100644
--- a/pkg/kubelet/dockershim/docker_streaming.go
+++ b/pkg/kubelet/dockershim/docker_streaming.go
@@ -21,6 +21,7 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"k8s.io/klog"
 	"math"
 	"time"
 
@@ -95,6 +96,8 @@ func (ds *dockerService) ExecSync(_ context.Context, req *runtimeapi.ExecSyncReq
 
 		exitCode = int32(exitError.ExitStatus())
 	}
+	klog.V(4).Infof(">>>> dockerService response size : %d", len(stdoutBuffer.Bytes()))
+	klog.V(4).Infof(">>>> dockerService response text : %s", string(stdoutBuffer.Bytes()))
 	return &runtimeapi.ExecSyncResponse{
 		Stdout:   stdoutBuffer.Bytes(),
 		Stderr:   stderrBuffer.Bytes(),
diff --git a/pkg/kubelet/prober/prober.go b/pkg/kubelet/prober/prober.go
index dd94f1705b23..ba5d62266100 100644
--- a/pkg/kubelet/prober/prober.go
+++ b/pkg/kubelet/prober/prober.go
@@ -258,7 +258,10 @@ type execInContainer struct {
 
 func (pb *prober) newExecInContainer(container v1.Container, containerID kubecontainer.ContainerID, cmd []string, timeout time.Duration) exec.Cmd {
 	return &execInContainer{run: func() ([]byte, error) {
-		return pb.runner.RunInContainer(containerID, cmd, timeout)
+		bytes, err := pb.runner.RunInContainer(containerID, cmd, timeout)
+		klog.V(4).Infof(">>>> newExecInContainer response size : %d", len(bytes))
+		klog.V(4).Infof(">>>> newExecInContainer response text : %s", string(bytes))
+		return bytes, err
 	}}
 }
 
diff --git a/pkg/probe/exec/exec.go b/pkg/probe/exec/exec.go
index b8cfe0d2ad7e..6ad5e9e9e2d3 100644
--- a/pkg/probe/exec/exec.go
+++ b/pkg/probe/exec/exec.go
@@ -57,6 +57,8 @@ func (pr execProber) Probe(e exec.Cmd) (probe.Result, string, error) {
 	}
 	data := dataBuffer.Bytes()
 
+	klog.V(4).Infof(">>>> Exec probe response length: %d", len(data))
+	klog.V(4).Infof(">>>> Exec probe response text: %s", string(data))
 	klog.V(4).Infof("Exec probe response: %q", string(data))
 	if err != nil {
 		exit, ok := err.(exec.ExitError)

```

```
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-exec
spec:
  containers:
  - name: liveness
    #image: k8s.gcr.io/busybox
    image: ubuntu
    args:
    - /bin/sh
    - -c
    - echo ok >/tmp/health; sleep 600
    #- echo ok >/tmp/health; sleep 10; rm -rf /tmp/health; sleep 600
    livenessProbe:
      exec:
        command:
        - cat
        - /tmp/health
      initialDelaySeconds: 15
      failureThreshold: 1

```

**Comment by liggitt** [comment]
/milestone v1.17

this isn't release-blocking, but can be picked to a patch release after soaking in master a bit

**Comment by tallclair** [comment]
@dims Can you run a test with a probe using the `yes` command (or some other infinite outputting command) as in the original report?

**Comment by dims** [comment]
@tallclair trying that now (command "yes")

**Comment by dims** [comment]
@tallclair the PR is holding up with the scenario in the bug repo.

**Comment by tallclair** [comment]
/lgtm
/approve

We should follow up to cap all `CombinedOutput()` calls, but that's lower priority.

**Pull request merged**